### PR TITLE
Changes to ElasticsearchService for pagination

### PR DESCRIPTION
**Pull request by alexwlchan**
## What is this PR trying to achieve?

Make the changes to the ElasticsearchService for pagination. Breaks up #275.

## Who is this change for?

Developers working on #275.

## Have the following been considered/are they needed?

- [x] Tests?
- [x] Docs?
- [x] Spoken to the right people?

### CODE REVIEW

File: api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticSearchService.scala [hunk 1]
```diff
@@ -24,21 +24,27 @@ class ElasticSearchService @Inject()(@Flag("es.index") index: String,
         get(id).from(s"$index/$itemType")
       }
 
-  def findResults(sortByField: String, limit: Int = 10): Future[RichSearchResponse] =
+  def listResults(sortByField: String,
```
Comment (alexwlchan) [inline api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticSearchService.scala]: Rename is per a comment in #275.

File: api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala [hunk 2]
```diff
@@ -47,40 +47,96 @@ class ElasticsearchServiceTest
...
+      expectedWorks = displayWorks.slice(5, 10)
+    )
+
+    // Take a slice that is beyond all our results
+    assertSliceIsCorrect(
+      searchService,
+      limit = 10,
+      from = 50,
+      expectedWorks = List()
+    )
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: @alicefuzier Hopefully easier to add more tests now. Is this the sort of thing you had in mind?

Comment (alicefuzier) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: No really, I just meant that in general the pagination functionality only had one test but I wasn't referring necessarily to this bit (sorry actually I was a bit unclear, but that is why I think it's better to go though pull requests in person)

Anyway, I like more testing, especially corner cases, but each assertion should be on its own test with a test name rather then comments (or in addition to comments if they are not redundant at that point)

**Comment by alexwlchan** [comment]
> Generally each test should test only one thing

Agreed, although we pay a cost in terms of reindexing everything to Elasticsearch, don't we? Still, change made.

I don't suppose ScalaTest has a way I can remove the repetition of `val displayWorks = populateElasticsearch()`, does it? We have fixtures in Python:

```
@pytest.fixture
def display_works(limit=10):
    return populate_elasticsearch(limit)

def test_foo(display_works):
    # the display_works variable is now available for use
```

File: api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala [hunk 3]
```diff
@@ -46,41 +46,86 @@ class ElasticsearchServiceTest
...
+   *  `DisplayWork` instances.
+   */
+  private def populateElasticsearch(count: Int = 10): List[DisplayWork] = {
+    val works: List[IdentifiedWork] = (0 to 9).map { x =>
+      identifiedWorkWith(canonicalId = s"ID-$x", label=s"Work number $x")
+    }.toList
+    insertIntoElasticSearch(works: _*)
+
+    works.map { work: IdentifiedWork =>
+      DisplayWork("Work", work.canonicalId, work.work.label)
```
Comment (alicefuzier) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: Just one quick question: why do we map to ```DisplayWork```? These are the tests for ```ElasticSearchService``` which should be agnostic about the fact that they are ```Work```s. Is there a way to make the assertions without assuming  anything about their structure? Maybe just reading them as strings?

File: api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala [hunk 4]
```diff
@@ -46,41 +46,86 @@ class ElasticsearchServiceTest
...
+    val searchResultFuture = searchService.listResults(
       sortByField = "canonicalId",
-      limit = 10
+      limit = limit,
+      from = from
     )
-    whenReady(searchResultFutureWithLargeLimit) { _.hits should have size 5 }
+    whenReady(searchResultFuture) { result =>
+      result.hits should have size expectedWorks.length
+      val returnedWorks = result.hits.map { DisplayWork(_) }
```
Comment (alicefuzier) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: same as above

File: api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala (hunk 3 above)
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: Convenience, mainly. That’s what the tests already had. I agree that these should be agnostic about the type of object they’re dealing with, but we have `Work` running throughout this setup.

I’d have no objection to pulling out the dependency on Work at a later date, but I don't fancy unpicking all those assumptions right now.

Comment (alicefuzier) [inline api/src/test/scala/uk/ac/wellcome/platform/api/services/ElasticsearchServiceTest.scala]: Makes sense 👍

**Comment by alicefuzier** [comment]
BTW forgot to reply to your question about repetition at the beginning of each test: you could use ```BeforeAndAfterEach``` trait and then override ```beforeEach`` to wrap setup code for each test. It's already in use in a lot of places so if you grep for it you should find examples
